Use webmozart/path-util for all path operations

m-vo opened the pull request:
### Applied transformations

**Joins**
* `'$a/'.$b.'/c'` / `sprintf('%s/%s/c', $a, $b)` &rarr; `Path::join($a, $b, 'c')`

**Parent**
* `dirname($a)` &rarr; `Path::getDirectory($a)`
* `dirname($a).'/b'` &rarr; `Path::join($a, '../b')`

**Normalization**
* `strtr($a, '\\', '/')` &rarr; `Path::normalize($a)`
* simplified `Configuration#canonicalize()` by reusing `Path::canonicalize()`

**Absolute/relative paths**
* `$filesystem->makePathRelative($path, $root)` &rarr; `Path::makeRelative($path, $root)`
* `str_replace(strtr($root, '\\', '/').'/', '', strtr($path, '\\', '/'))` &rarr; `Path::makeRelative($path, $root)`
* `substr($path, \strlen($root) + 1)` &rarr; `Path::makeRelative($path, $root)`
* `0 === strncmp($p, 'system/modules/', 15)` &rarr; `Path::isBasePath('system/modules', $p)`
* `0 === strncmp($path, $dir.'/', \strlen($dir) + 1)` &rarr; `Path::isBasePath($dir, $path)`
* `0 === strpos($path, $dir.'/')` &rarr; `Path::isBasePath($dir, $path)`
* `$filesystem->isAbsolutePath($path)` &rarr; `Path::isAbsolute($path)` 

**File extension**
* `pathinfo($filename, PATHINFO_EXTENSION)` &rarr; `Path::getExtension($resource, true)` (also handles uppercase file ext. if not explicitly wrapping with `strtolower()`)
* `'.twig' === substr($path, -5)` &rarr; `'twig' === Path::getExtension($path, true)`

**General**
* no more `rtim($path, '/')` (joined paths are already canonicalized)
* I introduced variables wherever the same path was used more than once.
* Some bits are simplified (e.g. `SymlinksCommand#filterNestedPaths`) now that we can use the helpers.

## Scope

All non-legacy code excluding tests.

*Note:* Concatenated paths includung  `__DIR__ ` are left untouched (see https://github.com/contao/contao/pull/1902#discussion_r449650482).

m-vo commented on core-bundle/tests/DependencyInjection/ConfigurationTest.php:
```diff
@@ -72,7 +72,9 @@ public function testResolvesThePaths(string $unix, string $windows): void
         $this->assertSame('/tmp/contao', $configuration['web_dir']);
-        $this->assertSame('C:\Temp\contao', $configuration['image']['target_dir']);
+
+        // todo: @fritzmg: this is what we talked about - not sure this is correct
```
ping @fritzmg

m-vo replied:
Nevermind, I'll leave this as is for now.

leofeyer commented on calendar-bundle/src/DependencyInjection/ContaoCalendarExtension.php:
```diff
@@ -16,14 +16,15 @@
         $loader = new YamlFileLoader(
             $container,
-            new FileLocator(__DIR__.'/../Resources/config')
+            new FileLocator(Path::join(__DIR__, '../Resources/config'))
```
I can click on `__DIR__.'/../Resources/config'` in my IDE and it will take me to the target. This does not work anymore with `Path::join()`, therefore I think we should not apply it to resolvable paths (e.g. starting with `__DIR__`).

m-vo replied:
Agreed. I left all these out for now.

As a sidenote: When writing new code, we should imo always consider canonicalizing paths that contain things like `__DIR__` or `__FILE__` (`Path::canonicalize(__DIR__ . '/../some/where')`) if they aren't immediately consumed by something that can surely handle them. 

This way things like `/../`, `a\b/c`, `a//b//c/`, ... won't ever appear inside a path &rarr; makes debugging and testing easier and prevents subtle mistakes. :slightly_smiling_face:

m-vo commented:
Scope of the changes now are `core-bundle/src` excluding `/Resource`. 

* There are some candidates in the `manager-bundle` and the `installation-bundle` as well. Should we include those or leave them out for now?
* Should we leave tests as they are for now? I think this might be a good idea as chances to break something by not noticing it (because the same transformation was applied to the source *and* tests) is smaller.
* I intentionally left out the legacy code. There are lots of candidates but I don't think it's worth it.

leofeyer commented:
> * There are some candidates in the `manager-bundle` and the `installation-bundle` as well. Should we include those or leave them out for now?
> * Should we leave tests as they are for now? I think this might be a good idea as chances to break something by not noticing it (because the same transformation was applied to the source _and_ tests) is smaller.
> * I intentionally left out the legacy code. There are lots of candidates but I don't think it's worth it.

Please adjust everything in all bundles except tests and legacy code. As you have said, the test suite is a good indicator to detect issues and we can adjust it in a second step.

m-vo committed:
use webmozart/path-util for path operations

m-vo committed:
add webmozart/path-util as a dependency

m-vo commented:
Allright, this PR is ready for review now. I adjusted the description accordingly.

Currently there are still some failing tests on windows that need to be taken care of. Imo the tests are wrong, not the code as they expect paths with mixed directory separators like so (which seems weird):

```
C:\here\is/a/path
```

Imo they should expect this instead (forward slashes no matter what):

```
C:/here/is/a/path
```

I also think `ContaoTestCase#getTempDir()` (see [here](https://github.com/contao/test-case/blob/b8ee4cb1a321cf1f7bcde8ee6c50acfb73370534/src/ContaoTestCase.php#L59-L74)) should return a normalized path instead of using the `DIRECTORY_SEPARATOR`. 

Wdyt?

@fritzmg: Could you check if my assumption is correct and maybe test the PR on windows as well? :slightly_smiling_face:

m-vo commented:
On a sidenote: We might want to homogenize the usage of `$filesystem->exists()` vs `file_exist()`. There are all kinds of variants and I left them untouched if not having to alter the line anyway (than I did use the `Filesystem` variant if already injected).

leofeyer commented on core-bundle/src/Cache/ContaoCacheWarmer.php:
```diff
@@ -155,12 +159,16 @@ private function generateLanguageCache(string $cacheDir): void
                     $dumper->dump(
                         iterator_to_array($subfiles),
-                        sprintf('languages/%s/%s.php', $language, $name),
+                        Path::join('languages', $language, "$name.php"),
```
```suggestion
                        Path::join('languages', $language, $name.'.php'),
```

m-vo replied:
Is there any reason why you'd want to have concatenation instead of interpolation? 
(besides personal preference which is fine oc :smile: )

leofeyer replied:
Concatenation is more performant than interpolation.

m-vo replied:
That's not true anymore for PHP7.

see https://blog.blackfire.io/php-7-performance-improvements-encapsed-strings-optimization.html

leofeyer replied:
Interesting. Based on this information, we should even prefer interpolation over concatenation?

m-vo replied:
Yes, I think so. But imo readability should come first as the performance difference is negligible in most cases.

leofeyer commented on core-bundle/src/DependencyInjection/Configuration.php:
```diff
@@ -422,38 +423,10 @@ static function (array $uris): bool {
-        }
-
-        return rtrim(implode('', $resolved), '\/');
+        return $windowsPath ? str_replace('/', '\\', $path) : $path;
```
```suggestion
        return $windowsPath ? strtr($path, '/', '\\') : $path;
```

m-vo replied:
see https://github.com/contao/contao/pull/1902#discussion_r450233694

leofeyer commented on core-bundle/src/DependencyInjection/Configuration.php:
```diff
@@ -422,38 +423,10 @@ static function (array $uris): bool {
     {
-        $resolved = [];
-        $chunks = preg_split('#([\\\\/]+)#', $value, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);
+        $windowsPath = false !== strpos($value, '\\');
```
Not sure if we should use this as an indicator. Why not `'\\' === \DIRECTORY_SEPARATOR`?

m-vo replied:
Because then you would not be able to unit test it.

ausi replied:
We could also remove the windows handling completeley here IMO, see https://github.com/contao/contao/pull/1902#discussion_r450095692

leofeyer replied:
We are running the tests on both Unix and Windows, so this should not be problematic to test.

m-vo replied:
No, I mean: if you depend on the `DIRECTORY_SEPARATOR`, then your unit tests have to do so as well. Otherwise you could not run a unit test with a 'Windows path' on Unix and expect it to output the right thing.

leofeyer replied:
Yes, of course, the unit test needs to use `DIRECTORY_SEPARATOR` as well when paths are not normalized.

m-vo replied:
see https://github.com/contao/contao/pull/1902#discussion_r450233694

leofeyer commented on core-bundle/src/DependencyInjection/Configuration.php:
```diff
@@ -463,13 +436,13 @@ private function getLocales(): array
-        if (is_dir($this->projectDir.'/contao/languages')) {
-            $dirs[] = $this->projectDir.'/contao/languages';
+        if (is_dir(Path::join($this->projectDir, 'contao/languages'))) {
+            $dirs[] = Path::join($this->projectDir, 'contao/languages');
```
```suggestion
        if (is_dir($path = Path::join($this->projectDir, 'contao/languages'))) {
            $dirs[] = $path;
```

I haven't tested this suggestion! It is just a reminder because your wrote
> I introduced variables wherever the same path was used more than once.

The same applies below.

ausi commented on core-bundle/src/Config/Loader/XliffFileLoader.php:
```diff
@@ -42,14 +43,14 @@ public function load($file, $type = null): string
         $xml = $this->getDomDocumentFromFile($name);
 
-        $return = "\n// ".str_replace(strtr($this->rootDir, '\\', '/').'/', '', strtr($name, '\\', '/'))."\n";
+        $return = "\n// ".str_replace(Path::normalize($this->rootDir).'/', '', Path::normalize($name))."\n";
```
```suggestion
        $return = "\n// ".Path::makeRelative($name, $this->rootDir)."\n";
```
If I’m understanding the code correctly?

m-vo replied:
Ah, yes of course.

leofeyer commented on core-bundle/src/Util/SymlinkUtil.php:
```diff
@@ -51,6 +47,9 @@ public static function symlink(string $target, string $link, string $rootDir): v
     public static function validateSymlink(string $target, string $link, string $rootDir): void
     {
+        $target = Path::normalize($target);
+        $link = Path::normalize($link);
```
We could move this after the `if ('' === $target) {` condition so we don't have to evaluate it if the first condition matches.

m-vo replied:
Done in 2fcc0ea72ce47cac179639ed9de6b7bb9d7def89.

leofeyer commented on installation-bundle/src/Config/ParameterDumper.php:
```diff
@@ -34,14 +35,18 @@ class ParameterDumper
-            $this->configFile = $rootDir.'/app/config/parameters.yml';
-        } elseif (!file_exists($this->configFile)) {
-            return;
+        if (!$this->filesystem->exists($this->configFile)) {
```
Why did you change the logic here?

ausi replied:
If I understand it correctly the `file_exists` for `$this->configFile` now only gets executed once instead of twice.

m-vo replied:
Before the paths were built 4 times, now twice. The logic is the same but the fallback is not checked prematurely resulting in less `file_exist` checks for some cases:

| default exists | fallback exists | =>       | checks old code                       | checks new code                       |
|----------------|-----------------|----------|---------------------------------------|---------------------------------------|
| yes            | yes             | default  | 2x exist(default), 1x exist(fallback) | 1x exist(default)                     |
| yes            | no              | default  | 1x exist(default), 1x exist(fallback) | 1x exist(default)                     |
| no             | yes             | fallback | 1x exist(default), 1x exist(fallback) | 1x exist(default), 1x exist(fallback) |
| no             | no              | -        | 1x exist(default), 1x exist(fallback) | 1x exist(default), 1x exist(fallback) |

leofeyer replied:
Do we have a unit test to verify that the method still behaves the same?

m-vo replied:
I added one.

* Here is the proof: 7fa64ae4dd400c5f7c4a2513613e7642b1c6a5e7 (I just replaced `file_exist()` with `$filesystem->exists()`.)
* Here is the new version again: a11de745851b55acbc69368aeabb9be5eb98bfe2.

leofeyer commented on manager-bundle/src/Command/DebugPluginsCommand.php:
```diff
@@ -131,16 +131,16 @@ private function listBundles(): int
             }
 
-            $rows[] = [$bundle->getName(), $path];
+            $rows[] = [$bundle->getName(), $path.'/'];
```
Why the trailing slash?

m-vo replied:
Because currently paths in the console output are listed with a trailing slash. We could also remove it + adjust the unit tests.

ausi replied:
I’d vote for removing it.

leofeyer replied:
Me too.

m-vo replied:
Changed in c235e49481be2565fd05dbcb60c9df0a853c5012.

leofeyer commented on manager-bundle/src/EventListener/InitializeApplicationListener.php:
```diff
@@ -35,14 +36,17 @@ public function __construct(string $projectDir)
     {
-        if (file_exists($this->projectDir.'/system/initialize.php')) {
+        $filesystem = new Filesystem();
+
```
```suggestion
```

ausi commented on core-bundle/src/DependencyInjection/Configuration.php:
```diff
@@ -422,38 +423,10 @@ static function (array $uris): bool {
-        }
-
-        return rtrim(implode('', $resolved), '\/');
+        return $windowsPath ? str_replace('/', '\\', $path) : $path;
```
Why do we need a special windows path handling here?

m-vo replied:
This is because - judging from the tests - a windows path like `C:\a\b\c` should keep it's slashes. 

I'm not 100% if this is needed. If not we could drop the custom canonicalization altogether.

ausi replied:
@contao/developers anyone an idea why we want windows paths with backslashes here?

m-vo replied:
This line: 
https://github.com/contao/contao/blob/d2eb923c90ec9b26fc69579309b89a334933490c/core-bundle/tests/DependencyInjection/ConfigurationTest.php#L75

would then have to be:

```php
        $this->assertSame('C:/Temp/contao', $configuration['image']['target_dir']);
```

ausi replied:
@leofeyer what do you think about removing the special windows handling here?

leofeyer replied:
If it works, I am all for it.

m-vo replied:
Adjusted in 24145736b7f726f9ccaf41dfa8f51f0b1af953e0.

leofeyer commented on manager-bundle/src/EventListener/InitializeApplicationListener.php:
```diff
@@ -35,14 +36,17 @@ public function __construct(string $projectDir)
 
-        (new Filesystem())
-            ->copy(
+        $filesystem->copy(
```
This is now a lot shorter and possibly fits in one line.

ausi commented on manager-bundle/src/Command/DebugPluginsCommand.php:
```diff
@@ -131,16 +131,16 @@ private function listBundles(): int
             }
 
-            $rows[] = [$bundle->getName(), $path];
+            $rows[] = [$bundle->getName(), $path.'/'];
```
Can’t we omit the trailing slash here?

m-vo replied:
see https://github.com/contao/contao/pull/1902#discussion_r450117526

ausi commented on manager-bundle/src/HttpKernel/ContaoKernel.php:
```diff
@@ -319,17 +320,21 @@ private function getConfigFile(string $file): ?string
+            $path = Path::join($rootDir, 'config', $file.$ext);
+
+            if (file_exists($path)) {
+                return $path;
```
```suggestion
            if (file_exists($path = Path::join($rootDir, 'config', $file.$ext))) {
                return $path;
```
Just a minor nitpick 🤓

m-vo replied:
I had this done already but got a (false positive) linter error and then did not find an easy way to work around. :roll_eyes: 

```
 Useless variable $path.

 Reported by:
"SlevomatCodingStandard\Sniffs\Variables\UselessVariableSniff.UselessVariable"
```

I changed it in 1d3a3991f4983c8fbbcfa3484e4d7d6fed511948 for you to see :slightly_smiling_face:.

ausi replied:
We should disable or fix this linter then IMO. But I’m also OK with leaving it as is...

leofeyer replied:
See https://github.com/slevomat/coding-standard/pull/1053

ausi commented on manager-bundle/src/HttpKernel/ContaoKernel.php:
```diff
@@ -319,17 +320,21 @@ private function getConfigFile(string $file): ?string
         // Fallback to the legacy config file (see #566)
         foreach (['.yaml', '.yml'] as $ext) {
-            if (file_exists($rootDir.'/app/config/'.$file.$ext)) {
+            $path = Path::join($rootDir, 'app/config', $file.$ext);
+
+            if (file_exists($path)) {
```
```suggestion
            if (file_exists($path = Path::join($rootDir, 'app/config', $file.$ext))) {
```

m-vo replied:
see https://github.com/contao/contao/pull/1902#discussion_r450123027

leofeyer replied:
We had to revert the changes for now due to https://github.com/slevomat/coding-standard/pull/1053.

ausi commented on manager-bundle/src/Routing/RouteLoader.php:
```diff
@@ -93,21 +94,25 @@ function (RouteCollection $collection, RoutingPluginInterface $plugin): RouteCol
         foreach (['routes.yaml', 'routes.yml', 'routing.yaml', 'routing.yml'] as $file) {
-            if (file_exists($this->rootDir.'/config/'.$file)) {
-                if (0 === strncmp($file, 'routing.', 8)) {
+            $path = Path::join($this->rootDir, 'config', $file);
+
+            if (file_exists($path)) {
```
```suggestion
            if (file_exists($path = Path::join($this->rootDir, 'config', $file))) {
```

m-vo replied:
see https://github.com/contao/contao/pull/1902#discussion_r450123027

ausi commented on manager-bundle/src/Routing/RouteLoader.php:
```diff
@@ -93,21 +94,25 @@ function (RouteCollection $collection, RoutingPluginInterface $plugin): RouteCol
         // Fallback to the legacy config file (see #566)
         foreach (['routes.yaml', 'routes.yml', 'routing.yaml', 'routing.yml'] as $file) {
-            if (file_exists($this->rootDir.'/app/config/'.$file)) {
+            $path = Path::join($this->rootDir, 'app/config', $file);
+
+            if (file_exists($path)) {
```
```suggestion
            if (file_exists($path = Path::join($this->rootDir, 'app/config', $file))) {
```

m-vo replied:
see https://github.com/contao/contao/pull/1902#discussion_r450123027

m-vo committed:
remove redundant path operation

Co-authored-by: Leo Feyer <github@contao.org>

m-vo committed:
remove empty line

Co-authored-by: Leo Feyer <github@contao.org>

m-vo committed:
inline variable definitions

m-vo committed:
collapse expression into one line

m-vo committed:
replace manual path normalization

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

m-vo committed:
optimize check

m-vo commented on core-bundle/src/Util/SymlinkUtil.php:
```diff
@@ -59,13 +55,15 @@ public static function validateSymlink(string $target, string $link, string $roo
+        $link = Path::normalize($link);
+
         if (false !== strpos($link, '../')) {
             throw new \InvalidArgumentException('The symlink path must not be relative.');
```
Sidenote: from a technical standpoint it would not be of any problem if the symlink path would include a ('../') - I assume we do this to prevent escaping the project dir level?

leofeyer replied:
Windows does not support relative symlinks. (It actually does, but only in a very limited scope.)

m-vo replied:
For windows `Filesystem#symlink` would receive a canonicalized absolute path now, so the relative links would be converted:
```
/path/to/somewhere/../foo  -->  /path/to/foo
```

leofeyer replied:
Since we are not sure why the code is there, I would highly recommend to leave it as is, unless

* you are 100% sure that you are not breaking anything and
* the change is technically required.

m-vo replied:
Sure, we need to keep it anyways because `validateSymlink()` is part of the public API...

m-vo committed:
remove redundant path operation

leofeyer commented on core-bundle/src/Util/SymlinkUtil.php:
```diff
@@ -26,20 +27,15 @@ public static function symlink(string $target, string $link, string $rootDir): v
-        if (!$fs->isAbsolutePath($target)) {
-            $target = $rootDir.'/'.$target;
-        }
+        $target = Path::makeAbsolute($target, $rootDir);
```
What if `$target` is an absolute path already?

m-vo replied:
Then it's returned unaltered.

m-vo committed:
remove trailing slashes in 'debug:plugins --bundles' command listing

m-vo committed:
add test for ParameterDumper

m-vo committed:
revert logic

m-vo committed:
fix additional whitespace

m-vo committed:
fix ContaoContextTest

m-vo committed:
adjust configuration for windows

m-vo committed:
fix tests on windows

-- this can be reverted once `TestCase#getTempDir()` and `TestCase#getFixturesDir()` return normalized paths

m-vo committed:
fix invalid test (does not fail without slash but because the path equals the root)

m-vo committed:
fix StringUtil::stripRootDir to allow paths with mixed directory separators

m-vo committed:
fix missed test (windows compat)

leofeyer commented on core-bundle/tests/DependencyInjection/Compiler/AddAssetsPackagesPassTest.php:
```diff
@@ -126,7 +127,7 @@ public function testUsesTheDefaultVersionStrategyForBundles(): void
     public function testUsesTheJsonManifestVersionStrategyForBundles(): void
     {
-        $bundlePath = static::getTempDir().'/ManifestJsonBundle';
+        $bundlePath = Path::normalize(static::getTempDir()).'/ManifestJsonBundle';
```
No `Path::join()` here?

m-vo replied:
I did not use it anywhere in the tests for now.

leofeyer replied:
Ok.

m-vo commented:
As a precaution I adjusted one part in the legacy code now - namely `StringUtil::stripRootDir()` - as this function previously did not work with paths containing mixed directory separators. I noticed this when adjusting tests (and could probably have worked around it) but I think this hardening is worth it anyway.

leofeyer commented on core-bundle/src/Resources/contao/library/Contao/StringUtil.php:
```diff
@@ -1159,7 +1160,7 @@ public static function stripRootDir($path)
 		$rootDir = System::getContainer()->getParameter('kernel.project_dir');
 		$length = \strlen($rootDir);
```
We should remove the `$length` variable and use `\strlen($rootDir)` in line 1168.

m-vo replied:
Done in cede8c8bf71643244f07f0325b8d8af553eabf5f.

m-vo committed:
inline variable

leofeyer commented on installation-bundle/tests/Config/ParameterDumperTest.php:
```diff
@@ -0,0 +1,57 @@
+        $fixtureDir = Path::canonicalize(__DIR__.'/../Fixtures');
+
+        $filesystem = $this->createMock(Filesystem::class);
+
```
```suggestion
```

leofeyer commented on installation-bundle/tests/Config/ParameterDumperTest.php:
```diff
@@ -0,0 +1,57 @@
+        ;
+
+        $dumper = new ParameterDumper($fixtureDir, $filesystem);
+
```
```suggestion
```

leofeyer reviewed (changes requested):
A few minor nitpicks need to be changed, then we are good to go. And please remove the `manager-bundle/tests/Command/.editorconfig` file again. It was not required before, therefore it seems specific to your setup.

**Edit:** Trailing whitespace is stripped indeed, so I guess we need a solution here. If we decide to fix this in the `.editorconfig` file, it needs to be fixed in the existing file in the root folder, though. However, we actually want the trailing whitespace to be stripped in the file – just not inside the heredoc. 🤔

m-vo committed:
drop empty lines

Co-authored-by: Leo Feyer <github@contao.org>

fritzmg commented:
Made a quick test installation on my Windows system:

✅ Symlinks were generated correctly
✅ Newly created public folders in the file manager also got a correct symlink
✅ `text` content element with image and chosen image size displays correctly in the front end

m-vo committed:
move expected output from DebugPluginsCommandTest into files

m-vo committed:
Merge branch 'feature/path-util-ftw' of github.com:m-vo/contao into feature/path-util-ftw

m-vo committed:
CS

leofeyer committed:
Work around https://github.com/slevomat/coding-standard/pull/1053

leofeyer committed:
Fix two minor issues in the YAML fixtures

leofeyer committed:
Merge branch 'master' into feature/path-util-ftw

leofeyer committed:
Fix the unit tests

leofeyer commented:
Thank you @m-vo.

aschempp commented:
Unfortunately, this seems to break when symlinking an extension in the vendor folder?

```
// /Users/aschempp/Projects/Contao/contao/core-bundle/src/Resources/contao/library/Contao/StringUtil.php

Path "/Applications/MAMP/htdocs/contao_master/../../../../Users/aschempp/Projects/Contao/contao/core-bundle/src/Resources/contao/templates/modules/mod_html.html5" is not inside the Contao root dir "/Applications/MAMP/htdocs/contao_master"
```

fritzmg commented:
@aschempp afaik that was already the case before. See https://github.com/contao/core-bundle/issues/1114

However, I wanted to look at that problem again, since some users stumble upon this problem, especially since I documented it this way in our dev docs 😬

m-vo commented:
~~That can easily be solved by using `Path::makeRelative()` instead of `substr()` in `StringUtil#stripRootDir()`. I left this untouched because people could have been relying on it working like this...~~

Edit: now I see, it's because the relative path goes up until the root... :thinking: 

Where do you encounter this error? (I don't get why this should be any different than before)

aschempp commented:
I symlinked `vendor/contao/contao` into my dev-master test application. Now I can't open the front end anymore. This setup worked fine before this PR.

fritzmg commented:
But this shouldn't have worked before either 🤔  (see my original issue about it)

m-vo commented:
@fritzmg Let me check if the old StringUtil implementation did allow this special case (which could be). One second.

fritzmg commented:
But that's what I am saying... it shouldn't have allowed it. Unless it allowed it after Contao 4.4.x.

m-vo commented:
@aschempp Do you have a call stack? Because the path *is not* inside the root dir and this check is imo wrong.

The old implementation apparently did allow this. :cry: 

```php
$path = '/a/b/../../../x'
$root = '/a/b'

// old code: '../../../x'
// new code: throws
```

m-vo commented:
We should probably just normalize the paths instead and keep the old implementation:
```diff

    static function stripRootDir($path)
    {
+       $path = Path::normalize($path);
+       $rootDir = Path::normalize(System::getContainer()->getParameter('kernel.project_dir'));

        $length = \strlen($rootDir);

        if (strncmp($path, $rootDir, $length) !== 0 || \strlen($path) <= $length || ($path[$length] !== '/' && $path[$length] !== '\\')) {
            throw new \InvalidArgumentException(sprintf('Path "%s" is not inside the Contao root dir "%s"', $path, $rootDir));
        }
    }
```

fritzmg commented:
But the old implementation does not allow it either.

Contao **4.4.49**:

```
InvalidArgumentException:
Path "C:\Users\fritzmg\vcs\contao-newslist-date-groups\src/Resources/contao/templates\news_latest_day.html5" is not inside the Contao root dir

  at vendor\contao\contao\core-bundle\src\Resources\contao\library\Contao\StringUtil.php:1100
  at Contao\StringUtil::stripRootDir('C:\\Users\\fritzmg\\vcs\\contao-newslist-date-groups\\src/Resources/contao/templates\\news_latest_day.html5')
     (vendor\contao\contao\core-bundle\src\Resources\contao\dca\tl_templates.php:234)
  at tl_templates->addNewTemplate(object(DC_Folder))
     (vendor\contao\contao\core-bundle\src\Resources\contao\classes\Backend.php:431)
  at Contao\Backend->getBackendModule('tpl_editor', null)
     (vendor\contao\contao\core-bundle\src\Resources\contao\controllers\BackendMain.php:131)
  at Contao\BackendMain->run()
     (vendor\contao\contao\core-bundle\src\Controller\BackendController.php:55)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor\symfony\symfony\src\Symfony\Component\HttpKernel\HttpKernel.php:151)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor\symfony\symfony\src\Symfony\Component\HttpKernel\HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor\symfony\symfony\src\Symfony\Component\HttpKernel\Kernel.php:200)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web\app_dev.php:67)
```

Contao **4.9.3**:

```
InvalidArgumentException:
Path "C:\Users\fritzmg\vcs\contao-newslist-date-groups\src/Resources/contao/templates\news_latest_day.html5" is not inside the Contao root dir "C:\Users\fritzmg\www\c49dev"

  at vendor/contao/contao/core-bundle/src/Resources/contao/library/Contao/StringUtil.php:1118
  at Contao\StringUtil::stripRootDir('C:\\Users\\fritzmg\\vcs\\contao-newslist-date-groups\\src/Resources/contao/templates\\news_latest_day.html5')
     (vendor/contao/contao/core-bundle/src/Resources/contao/dca/tl_templates.php:241)
  at tl_templates->addNewTemplate(object(DC_Folder))
     (vendor/contao/contao/core-bundle/src/Resources/contao/classes/Backend.php:441)
  at Contao\Backend->getBackendModule('tpl_editor', null)
     (vendor/contao/contao/core-bundle/src/Resources/contao/controllers/BackendMain.php:167)
  at Contao\BackendMain->run()
     (vendor/contao/contao/core-bundle/src/Controller/BackendController.php:48)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor/symfony/http-kernel/HttpKernel.php:158)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:80)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:201)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/index.php:31)
  at require('C:\\Users\\fritzmg\\www\\c49dev\\web\\index.php')
     (web/app.php:4)
```

m-vo commented:
see https://github.com/contao/contao/pull/1918
